### PR TITLE
Loosen `linalg_ext.scatter` requirements for partial update

### LINKED ISSUE #7628: "'linalg_ext.scatter' op mismatch in shape of update value dim#1 and original value at dim#1" for `mhlo.scatter` lowering
Repro:
```
from iree.compiler import compile_str

CODE = """
module  {
  func @main(%arg0: tensor<10x5xcomplex<f32>>, %arg1: tensor<3x1xi32>, %arg2: tensor<3x3xcomplex<f32>>) -> tensor<10x5xcomplex<f32>> {
    %0 = "mhlo.scatter"(%arg0, %arg1, %arg2) ( {
    ^bb0(%arg3: tensor<complex<f32>>, %arg4: tensor<complex<f32>>):  // no predecessors
      %1 = mhlo.add %arg3, %arg4 : tensor<complex<f32>>
      "mhlo.return"(%1) : (tensor<complex<f32>>) -> ()
    }) {indices_are_sorted = false, scatter_dimension_numbers = #mhlo.scatter<update_window_dims = [1], inserted_window_dims = [0], scatter_dims_to_operand_dims = [0], index_vector_dim = 1>, unique_indices = false} : (tensor<10x5xcomplex<f32>>, tensor<3x1xi32>, tensor<3x3xcomplex<f32>>) -> tensor<10x5xcomplex<f32>>
    return %0 : tensor<10x5xcomplex<f32>>
  }
}
"""

compiled_flatbuffer = compile_str(CODE, target_backends=["dylib"], input_type="mhlo")
```

Get:
```
$ python t.py
Traceback (most recent call last):
  File "t.py", line 16, in <module>
    compiled_flatbuffer = compile_str(CODE, target_backends=["dylib"], input_type="mhlo")
  File "/usr/local/google/home/phawkins/.pyenv/versions/py373a/lib/python3.7/site-packages/iree/compiler/tools/core.py", line 262, in compile_str
    result = invoke_immediate(cl, immediate_input=input_bytes)
  File "/usr/local/google/home/phawkins/.pyenv/versions/py373a/lib/python3.7/site-packages/iree/compiler/tools/binaries.py", line 201, in invoke_immediate
    raise CompilerToolError(process)
iree.compiler.tools.binaries.CompilerToolError: Error invoking IREE compiler tool ireec
Diagnostics:
<stdin>:4:10: error: 'linalg_ext.scatter' op mismatch in shape of update value dim#1 and original value at dim#1
    %0 = "mhlo.scatter"(%arg0, %arg1, %arg2) ( {
         ^
<stdin>:3:3: note: called from
  func @main(%arg0: tensor<10x5xcomplex<f32>>, %arg1: tensor<3x1xi32>, %arg2: tensor<3x3xcomplex<f32>>) -> tensor<10x5xcomplex<f32>> {
  ^
<stdin>:2:1: error: conversion from source -> vm failed
module  {
^


Invoked with:
 ireec /usr/local/google/home/phawkins/.pyenv/versions/py373a/lib/python3.7/site-packages/iree/compiler/tools/../_mlir_libs/ireec - --iree-input-type=mhlo --iree-vm-bytecode-module-output-format=flatbuffer-binary --iree-hal-target-backends=dylib --iree-mlir-to-vm-bytecode-module --iree-llvm-embedded-linker-path=/usr/local/google/home/phawkins/.pyenv/versions/py373a/lib/python3.7/site-packages/iree/compiler/tools/../_mlir_libs/iree-lld --mlir-print-debuginfo --mlir-print-op-on-diagnostic=false
```

### PR DESCRIPTION
**Pull request by rsuderman**
mhlo.scatter has implicit 0 index for the update and allows a partial update for along the destination axis.

### CODE REVIEW

File: compiler/src/iree/compiler/InputConversion/MHLO/test/convert_mhlo_to_linalg_ext.mlir [hunk 1]
```diff
@@ -343,6 +343,27 @@ func.func @scatter_add_slice_2D(%arg0: tensor<6x3xi32>, %arg1: tensor<2x1xi32>,
...
+  %0 = "mhlo.scatter"(%arg0, %arg1, %arg2) ( {
+  ^bb0(%arg3: tensor<f32>, %arg4: tensor<f32>):  // no predecessors
+    %1 = mhlo.add %arg3, %arg4 : tensor<f32>
+    "mhlo.return"(%1) : (tensor<f32>) -> ()
+  }) {indices_are_sorted = false, scatter_dimension_numbers = #mhlo.scatter<update_window_dims = [1], inserted_window_dims = [0], scatter_dims_to_operand_dims = [0], index_vector_dim = 1>, unique_indices = false} : (tensor<10x5xf32>, tensor<3x1xi32>, tensor<3x3xf32>) -> tensor<10x5xf32>
+  return %0 : tensor<10x5xf32>
+}
+
+// CHECK-LABEL: func.func @scatter_partial
+// CHECK:         %[[ARG0:[a-zA-Z0-9]+]]
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/test/convert_mhlo_to_linalg_ext.mlir:356]: Nit: CHECK-SAME

**Review by jpienaar — APPROVED** [review]
Do we have a numerical end-to-end test for this one too? (not blocking)

File: compiler/src/iree/compiler/InputConversion/MHLO/test/convert_mhlo_to_linalg_ext.mlir (hunk 1 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/MHLO/test/convert_mhlo_to_linalg_ext.mlir]: Done.

**Comment by rsuderman** [comment]
> Do we have a numerical end-to-end test for this one too? (not blocking)

Nope so I added one.

**Comment by MaheshRavishankar** [comment]
I was OOO, so just saw this. I am not sure whats happening here (its checked in, so thats fine), but something here tells me we might be pushing the boundaries of scatter than we should. Could you summarize the change (I cant follow from the PR)
